pandas/core/arrays/base.py

```diff
@@ -266,3 +266,15 @@ def _can_hold_na(self):
         Setting this to false will optimize some operations like fillna.
         """
         return True
+
+    @property
+    def _ndarray_values(self):
+        # type: () -> np.ndarray
+        """Internal pandas method for lossy conversion to a NumPy ndarray.
+
+        This method is not part of the pandas interface.
```

Comment:
I find it a bit strange that we say this is not part of the interface, but still provide here a default implementation and say what it is. 

If it is not part of the interface, we could also define it on our own subclasses without defining it here. 
But on the other hand, that will raise errors when an external extension array does not have this method. Eg some attributes on Series will call into this, which makes it somehow part of the interface.

Reply:
> I find it a bit strange that we say this is not part of the interface, but still provide here a default implementation and say what it is.

This simplifies the implementation since

a.) we don't have to have a `PandasExtensionArray` that subclasses `ExtensionArray`, just to define this method.
b.) we can safely call `._values` anywhere in our code without checking whether it's an extension array or a pandas extension array.

My preference is to leave it out of the interface until someone sees an actual need for it. I suspect this could come up if / when we start allowing custom indexes with their own indexing engines, but that seems like a ways down the road...

Reply:
Yes, I understand those reasons, but that means it *is* part of the interface (in the sense that if somebody would be stupid to implement a `_ndarray_values` property that returns something differently, it will break code?) 
It's just a part of the interface that we ask not to implement?

To give an example, for GeoPandas, I was wondering if I would overwrite this property to return my `self.data` (the integer pointers) instead of a materialized array (which can be costly). 
For things like `Series.strides`, `Series.itemsize`, .. this will be OK (and in principle no user should call this anyway .. so maybe I should not worry about it), but if it is used in other places as well that I am not really sure about, such an implementation might actually give a wrong result (but at the same time, if this is called in certain places for my column with an extension array, I want to know that, because materializing can be costly and I want to avoid this as much as possible).

Reply:
I see your point. Your GeoPandas concern is a good one. But I'm not sure how to proceeded :/ Do you have a preference for

a.) Adding it to the interface
b.) Leaving it out of the interface, with a default implementation on EA
c.) Making a pandas-specific EA that defines `_ndarray_values`, so uses of `_ndarray_values` will need to check for that attr.

`._values` / `._ndarray_values` are meant to be internal, and I don't want to limit future refactorings with backwards compatibility concerns. I'm also not even sure how to document it beyond what's already there.

Reply:
Maybe the best option is to leave it for now? :)

Ideally, the Series machinery should not use `_ndarray_values` too often, as this would require knowledge of what those values mean (codes, ordinals, ...), which Series does not need to know (if there is knowledge required, it should somehow dispatch the operation to the underlying (extension) array itself). 

And this is already the case I think. From a quick scan, currently in the PR, I see `ndarray_values` is mainly used for:

- indexing (values used in the engine) -> for now we don't support extension arrays to be the data for Index (apart from our own), so this is a worry for later if we want to tackle that
- Index object operations (eg the setops) -> same as above
- in some specific cases where we know the parent object (eg we know it is a PeriodIndex, we just want to get the ordinals)
- for algos (eg take, unique) -> either they dispatch to EA (eg take) or for now the fallback is object array anyhow (eg value_counts), and in general this is something we still need to discuss how to handle in general (factorize, unique, ..)
- some Series attributes (shape, itemsize, strides, size) -> shape can also be derived from `_values` directly (like `nbytes`) as it is part of the EA interface. And from the others, I think `size` would be nice to not have to materialize the EA (itemsize and strides I don't care about). Can we let this return `len(self._values)`, which is part of the interface? (I am not sure to what extent `.size` and `__len__` on numpy arrays are different / have different performance)

Reply:
Your assessment looks correct. Nothing that's series specific currently uses `_ndarray_values`, so it's really just code that is Index-specific or index / series agnostic.

I can change `.shape` and `.size` to use `._values` instead of  instead of `._ndarary` values. I'll just have to override it in `DatetimeIndex` to avoid a recursion error.